### PR TITLE
feat: negative impls

### LINKED ISSUE #652: negative_impls experimental features in scarb.toml (maybe not required in the book for now as it might change in the future

### PR DESCRIPTION
**Pull request by enitrat**
Closes #652

This change is [Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/650)

### CODE REVIEW

File: src/ch08-02-traits-in-cairo.md [hunk 1]
```diff
@@ -140,3 +140,22 @@ Implementations can be aliased when imported. This is most useful when you want
...
+
+## Negative impls
+
+> Note: This is still an experimental feature and can only be used if `experimental-features = ["negative_impls"]` is enabled in your Scarb.toml file, under the `[package]` section.
+
+Negative implementations, also known as negative traits or negative bounds, are a mechanism that allows you to express that a type does not implement a certain trait when defining the implementation of a trait over a generic type. Negative impls enable you to write implementations that are applicable only when another implementation does not exist in the current scope.
+
+For example, let's say we have a trait `Producer` and a trait `Consumer`, and we want to define a generic behavior where all types implement the `Consumer` by default. However, we want to ensure that no type can be both a `Consumer` and a `Producer`. We can use negative impls to express this restriction.
```
Comment (TAdev0) [inline src/ch08-02-traits-in-cairo.md:150]: the 'Consumer' **trait** by default

File: listings/ch08-generic-types-and-traits/no_listing_18_negative_impl/src/lib.cairo [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+trait Producer<T> {
+    fn produce(self: T) -> u32;
+}
+
+trait Consumer<T> {
+    fn consume(self: T, input: u32);
+}
+
+impl ProducerImpl of Producer<ProducerType> {
```
Comment (TAdev0) [inline listings/ch08-generic-types-and-traits/no_listing_18_negative_impl/src/lib.cairo:18]: following the previously discussed topic about method / trait impl , you ended up with this syntax of trait impl: 

```
trait Shape<T>{
fn area(self: @T) -> u32;
}

impl CircleShape of Shape<Circle> {}

impl RectangleShape of Shape<Rectangle>{}
```

this means this example should be : 
```
impl ProducerTypeProducer of Producer<ProducerType>
```

but this looks weird

File: listings/ch08-generic-types-and-traits/no_listing_18_negative_impl/src/lib.cairo [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+fn main() {
+    let producer = ProducerType {};
+    let another_type = AnotherType {};
+    let third_type = AThirdType {};
+    let production = producer.produce();
+
+    // producer.consumer(production); Invalid: ProducerType does not implement Consumer
+    another_type.consume(production);
+    third_type.consume(production);
+}
```
Comment (TAdev0) [inline listings/ch08-generic-types-and-traits/no_listing_18_negative_impl/src/lib.cairo:39]: really nice example overall , very clear!

File: listings/ch08-generic-types-and-traits/no_listing_18_negative_impl/src/lib.cairo (hunk 2 above)
Comment (enitrat) [reply listings/ch08-generic-types-and-traits/no_listing_18_negative_impl/src/lib.cairo:18]: I dont think it matters a lot what the implementation name is in that case (nor is it in the example you're writing - we're very rarely going to use the impl directly

**Review by TAdev0 — APPROVED** [review]
lgtm!

**Pull request merged**